packages/components/src/DataList/components/DataListLayoutInternal/DataListItems.tsx

```diff
@@ -19,41 +23,83 @@ export function DataListItems<T extends DataListObject>({
+  const isContextMenuVisible = Boolean(contextPosition);
 
   return (
     <DataListLayoutInternal
       layouts={layouts}
       mediaMatches={mediaMatches}
-      renderLayout={layout => {
-        return (
-          <>
-            {elementData.map((child, i) => {
-              const item = data[i];
-
-              return (
-                <div
-                  onMouseEnter={() => setHover(item.id)}
-                  onMouseLeave={() => setHover(undefined)}
-                  className={styles.listItem}
-                  key={item.id}
-                >
-                  <DataListItemInternal item={data[i]}>
-                    {layout.props.children(child)}
-                  </DataListItemInternal>
-
-                  <AnimatePresence>
-                    {hover === item.id && (
-                      <InternalDataListItemActions item={item} />
-                    )}
-                  </AnimatePresence>
-                </div>
-              );
-            })}
-          </>
-        );
-      }}
+      renderLayout={layout => (
+        <>
+          {elementData.map((child, i) => {
+            const item = data[i];
+            const isActive = activeID === item.id;
+            const shouldShowContextMenu =
+              isActive && isContextMenuVisible && Boolean(contextMenuActions);
+
+            return (
+              <div
+                onMouseEnter={handleMouseEnter(item)}
+                onMouseLeave={handleMouseLeave}
+                onContextMenu={handleContextMenu}
+                className={classNames(styles.listItem, {
+                  [styles.active]: isActive && isContextMenuVisible,
+                })}
+                key={item.id}
+              >
+                <DataListItemInternal item={data[i]}>
+                  {layout.props.children(child)}
+                </DataListItemInternal>
+
+                <AnimatePresence>
+                  {isActive && <InternalDataListItemActions item={item} />}
+
+                  <DataListActionsMenu
+                    key={item.id}
+                    visible={shouldShowContextMenu}
+                    position={contextPosition || { x: 0, y: 0 }}
+                    onRequestClose={() => setContextPosition(undefined)}
+                  >
+                    {contextMenuActions &&
+                      Children.map(contextMenuActions, action => (
+                        <InternalDataListAction
+                          key={item.id}
+                          {...action.props}
+                          item={item}
+                        />
+                      ))}
+                  </DataListActionsMenu>
+                </AnimatePresence>
+              </div>
+            );
+          })}
+        </>
+      )}
     />
   );
+
+  function handleMouseEnter(item: T) {
+    return () => setActiveID(item.id);
+  }
+
+  function handleMouseLeave() {
+    !contextPosition && setActiveID(undefined);
```

Comment:
That was autofixed. I started with Boolean first. haha

<img width="709" alt="image" src="https://github.com/GetJobber/atlantis/assets/15986172/5828ffd9-21b9-426a-b624-c020154081bf">

It probably should be a `!isContextMenuVisible` tho so good catch.

As for `undefined`. That argument is actually new to me since we are using Typescript and Atlantis is purely using `undefined` for things (like Typescript as well). The only use of `null` in Atlantis is from refs.

> bug happened and it's undefined

Can't think of an example where that would happen on logic where you're explicitly setting a state. I can see that on `functions` that should be returning a `string` but then returns `undefined` instead since that's how `void` works. But still, TS would catch that and throw an error about 

> This function is returning undefined but it should be returning a string

I see what you're saying tho since GQL and Apollo does require the difference between null and undefined.

Reply:
yeah fair, I mean TS _should_ be protecting us from code paths that could result in that. 

honestly I don't know enough about DataList at the moment to present any super informed cases where I'd reasonably expect an `undefined` could sneak in undetected and cause the vague scenario I mentioned

so I'm fine with not doing anything here

and yes, I agree about `null` vs `undefined` which could become a lot more relevant as we integrate with the backend which might have different, meaningful implications for `undefined` vs `null`